x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/QueryableBuiltInRolesSynchronizer.java

```diff
@@ -0,0 +1,488 @@
+        return true;
+    }
+
+    private void doSyncBuiltinRoles(Map<String, String> indexedRolesDigests, QueryableBuiltInRoles roles, ActionListener<Void> listener) {
+        securityIndex.prepareIndexIfNeededThenExecute(listener::onFailure, () -> {
+            final SecurityIndexManager frozenSecurityIndex = securityIndex.defensiveCopy();
+            if (frozenSecurityIndex.isAvailable(SecurityIndexManager.Availability.PRIMARY_SHARDS) == false) {
+                listener.onFailure(frozenSecurityIndex.getUnavailableReason(SecurityIndexManager.Availability.PRIMARY_SHARDS));
+
+            } else {
+                final Collection<RoleDescriptor> rolesToUpsert = rolesToUpsert(roles, indexedRolesDigests);
+                indexRoles(rolesToUpsert, frozenSecurityIndex, ActionListener.wrap(onResponse -> {
+                    final Set<String> rolesToDelete = rolesToDelete(roles, indexedRolesDigests);
+                    if (rolesToDelete.isEmpty()) {
+                        markRolesAsSynced(frozenSecurityIndex.getConcreteIndexName(), indexedRolesDigests, roles.rolesDigest(), listener);
+                    } else {
+                        deleteRoles(frozenSecurityIndex, roles, rolesToDelete, indexedRolesDigests, listener);
+                    }
+                }, listener::onFailure));
+            }
+        });
+    }
+
+    private static Set<String> rolesToDelete(QueryableBuiltInRoles roles, Map<String, String> indexedRolesDigests) {
+        return indexedRolesDigests == null ? Set.of() : Sets.difference(indexedRolesDigests.keySet(), roles.rolesDigest().keySet());
+    }
+
+    private static Collection<RoleDescriptor> rolesToUpsert(QueryableBuiltInRoles roles, Map<String, String> indexedRolesDigests) {
+        final Set<RoleDescriptor> rolesToUpsert = new HashSet<>();
+        for (var role : roles.roleDescriptors()) {
+            final String roleDigest = roles.rolesDigest().get(role.getName());
+            if (indexedRolesDigests == null || indexedRolesDigests.containsKey(role.getName()) == false) {
+                rolesToUpsert.add(role);  // a new role to create
+            } else if (indexedRolesDigests.get(role.getName()).equals(roleDigest) == false) {
+                rolesToUpsert.add(role);  // an existing role that needs to be updated
+            }
+        }
+        return rolesToUpsert;
+    }
+
+    private void deleteRoles(
+        SecurityIndexManager securityIndex,
+        QueryableBuiltInRoles roles,
+        Set<String> rolesToDelete,
+        Map<String, String> indexedRolesDigests,
+        ActionListener<Void> listener
+    ) {
+        queryableBuiltInRolesStore.deleteRoles(securityIndex, rolesToDelete, ActionListener.wrap(deleteResponse -> {
+            final Optional<Exception> deleteFailure = deleteResponse.getItems()
+                .stream()
+                .filter(BulkRolesResponse.Item::isFailed)
+                .map(BulkRolesResponse.Item::getCause)
+                .findAny();
```

Comment:
Correct. Currently, we will report back only a single failure and make a decision only based on single one. 
Alternative option I see is to create a dedicated exception which holds all failure causes. Then we can report and process all of them. Not sure at the moment if this is a good idea, but probably better that reacting and logging only a single failure.

Reply:
I've improved exception handling by introducing dedicated exceptions and handling all bulk failures (instead of reporting and handling a single one): https://github.com/elastic/elasticsearch/pull/117581/commits/fec25ae67b21051182e9f7912badf0b9aff7a68a